### DISCUSSION TITLE
How to properly add custom filetypes?

### DISCUSSION THREAD

**Discussion by ghost** [geany/geany#3663] [Q&A] [answered]
The language is based on Go. I created `filetypes.Foo.conf` and set `Foo=*.foo;` in `filetype_extensions.conf`. Foo is indeed appeared on `Set Filetype` but syntax highlighting for the language Foo is simply not work. When you opened a file name with extension `.foo`, it's treated as `None` and even if you manually set the language to be `Foo` via `Set Filetype`, there is absolutely no syntax highlighting. Strangely enough, if I set the language to `Go` via `Set Filetype` then I have syntax highlighting. If I edit `filetype_extensions.conf` to add `.foo` to the list of extensions of `Go` source file, then `.foo` files will be treated as `Go` and have syntax highlighting, too. Of course, this syntax highlighting is for Go and the keywords specific to the language `Foo` is not highlighted. I have read and followed the FAQ so I want to know what exactly did I do wrong or if I should make a bug report.

**Comment by elextr** [comment]
Did you set a lexer to use like [this](https://github.com/geany/geany/blob/ba6fc4a4fa48239daca7350ec05d25e44cc89169/data/filedefs/filetypes.Scala.conf#L16)?

It is hard to debug with no information, better to put your filetype file in a gist and paste a link here (don't paste here and spam every watcher's email).

**Reply by ghost** [reply]
Copy `filetypes.go` into `filetypes.Foo.conf`. Edit `filetypes.Foo.conf` to set `extension=foo` and `mime_type=text/x-foo`. Set `Foo=*.foo;` in `filetype_extensions.conf`. Copy a sample Go source file into `sample.foo` (or simply replace the extension from `.go` to `.foo`). Open `sample.foo` in Geany. This is the steps to reproduce.

**Comment by elextr** [comment, marked as answer]
Go is a builtin filetype, so `filetypes.go` cannot simply be copied as a custom filetype, thats why my link above pointed to a custom filetype file (Scala as an example).

**Reply by ghost** [reply]
OK. It's working now. But I want to ask you for sure that it should be `lexer_filetype=C` or not. I don't think Go is a C-style language.

**Reply by elextr** [reply]
Well it should be whichever of the builtin filetypes is closest to your language, from the sound of it that would be Go.
